File: tetris/rTetris/board.rb
require 'piece.rb'

class Board
  attr_accessor :tabuleiro, :peca, :next
	
  def nova_peca
    @next = Piece.new
  end

  def proximo
    @peca = @next
    nova_peca
  end
  
  def esquerda
    @peca.x -= 1
    @peca.x += 1 if colide?
  end
  
  def direita
    @peca.x += 1
    @peca.x -= 1 if colide?
  end
  
  def baixo
    if @mr
      @peca.y += 1
    else
      @peca.y += 0.1
    end
    		
    colidiu = colide?
		
    if colidiu or @peca.y + @peca.matriz.length > 21
      @peca.y = 21 - self.peca.matriz.length if @peca.y + @peca.matriz.length > 21
      fixar
    end
  end
	
  def mais_rapido(baixo)
    @mr = baixo
  end
	
  def colide?
    selfy = @peca.y.to_i
		
    @peca.matriz[0].length.times do |x|
      @peca.matriz.length.times do |y|
        begin
	  next if x+@peca.x < 0 or y + selfy < 0
	  if @peca.matriz[y][x] == 1 and @tabuleiro[y+selfy][x+@peca.x] == 1
  	    return true
	  end
	rescue
	end
      end
    end
    return false
  end
	
  #Trava a peça no tabuleiro
  def fixar
    @peca.y = @peca.y.to_i - 1
    for x in @peca.x...@peca.matriz[0].length + @peca.x
      for y in @peca.y...@peca.matriz.length + self.peca.y.to_i
        if x < 0 or y < 0
          fail "Game Over"
        elsif @peca.matriz[y-@peca.y.to_i][x-@peca.x] == 1
          @tabuleiro[y][x] = 1
        end
      end
    end
      	
    limpa
    proximo
  end
		
  #Rotaciona a peça.
  def rotacionar
    colunas = @peca.matriz[0].length
      	
    #Acertar as linhas e colunas:
    antx = self.peca.x

    @peca.x += colunas / 2 - @peca.matriz.length / 2
    if colide? or @peca.x + @peca.matriz.length > 10 or @peca.x < 0
      @peca.x = antx
      return
    end

    @peca.matriz = @peca.matriz.transpose.reverse
  end

  def limpa
    20.times do |y|
      if @tabuleiro[y] == [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]:
        @tabuleiro.delete_at y
        @tabuleiro = [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]] + @tabuleiro
      end
    end
  end

  def initialize
    @mr = false
    @tabuleiro = []
    20.times do |y|
      @tabuleiro += [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]
    end

    nova_peca
    proximo
  end
end

File: tetris/pyTetris/Board.py
# -*- coding: utf-8 -*-

import random
import time
from Piece import Piece

class Board:
    tabuleiro = []
    mais_rapido = False
    
    def nova_peca(self):
        self.next = Piece()
    
    def proximo(self):
        self.peca = self.next
        self.nova_peca()
    
    def esquerda(self):
        if self.peca.x == 0: return
        self.peca.x -= 1
        if(self.colide()): self.peca.x += 1
    
    def direita(self):
        if self.peca.x + len(self.peca.matriz[0]) == 10: return
        self.peca.x += 1
        if(self.colide()): self.peca.x -= 1
    
    def baixo(self):
        if self.mais_rapido: self.peca.y += 1
        else: self.peca.y += 0.1
        
        colidiu = self.colide()
        if colidiu or self.peca.y + len(self.peca.matriz) > 21:
            if self.peca.y + len(self.peca.matriz) > 21: self.peca.y = 21 - len(self.peca.matriz) 
            self.fixar()
    
    def colide(self):
        selfy = int(self.peca.y)
    
        for x in range(len(self.peca.matriz[0])):
            for y in range(len(self.peca.matriz)):
                try:
                    if x+self.peca.x < 0 or y+int(selfy) < 0: continue
                    if self.peca.matriz[y][x] == 1 and self.tabuleiro[y+int(selfy)][x+self.peca.x] == 1:
                        return True
                except:
                    pass
        return False
    
    def fixar(self):
        "Trava a peça no tabuleiro"
        self.peca.y = int(self.peca.y)  - 1
        for x in range(self.peca.x, len(self.peca.matriz[0]) + self.peca.x):
            for y in range(int(self.peca.y), len(self.peca.matriz) + int(self.peca.y)):
                if x < 0 or y < 0:
                    raise Exception("Game Over")
                if self.peca.matriz[y-int(self.peca.y)][x-self.peca.x] == 1:
                    self.tabuleiro[y][x] = 1
        
        self.limpar()
        self.proximo()
        
    def rotacionar(self):
        "Rotaciona a peça."
        colunas = len(self.peca.matriz[0])
        
        "Acertar as linhas e colunas:"
        antx = self.peca.x
        
        self.peca.x += colunas / 2 - len(self.peca.matriz) / 2
        if self.colide() or self.peca.x + len(self.peca.matriz) > 10 or self.peca.x < 0:
            self.peca.x = antx
            return
        
        self.peca.matriz = zip(*self.peca.matriz)
        self.peca.matriz.reverse()
    
    def limpar(self):
        for y in range(0, 20):
            if self.tabuleiro[y] == [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]:
                del(self.tabuleiro[y])
                self.tabuleiro.insert(0, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
    
    
    def __init__(self):
        random.seed = time.time()
        for y in range(20):
            self.tabuleiro.append( [0, 0, 0, 0, 0, 0, 0, 0, 0, 0] )
        
        self.nova_peca()
        self.proximo()
